content/posts/cicd_ha.md

```diff
@@ -0,0 +1,132 @@
+
+But not every use case or business requires the same level of HA. In other terms, the word *High* in the *HA* term is relative. A critical service interruption can be seconds, minutes or even hours depending on the context, use case or business.
+
+## What about HA and CI/CD?
+
+Continuous Integration and Continuous Delivery (CI/CD) is based on the concept of continuously developing, building, integrating, testing, delivering and releasing the software, with the end goal of producing high quality applications for demanding users. So, how should we define a highly available platform or solution to be able to do that with no *"service interruption"*?
+
+Let's see what a *service interruption* means in this case.
+
+### Industrial manufacturing, the example to look at
+
+I think that software development automation is taking all the learnings from the industrial manufacturing processes and its *Continuous Delivery* systems where final products tend to be manufactured on demand, saving a lot of unnecessary costs and making production lines a lot more agile with very high quality end user products (concepts like [JIT](https://en.wikipedia.org/wiki/Just-in-time_manufacturing), [Kanban](https://en.wikipedia.org/wiki/Kanban), [ConWIP](https://en.wikipedia.org/wiki/CONWIP), [Lean](https://www.sciencedirect.com/science/article/pii/S1877705814034092)... that are usually known for software development were born at the Industrial Manufacturing evolution during the 20th Century).
+
+This is exactly what CI/CD is doing for software development. An [assembly line](https://en.wikipedia.org/wiki/Assembly_line) is the exact same concept for industrial manufacturing as it is for a [CI/CD pipeline](https://dzone.com/articles/learn-how-to-setup-a-cicd-pipeline-from-scratch) for software development. Just an automated sequence of stages (can be parallel or not) that execute different steps to deliver high quality products in a frequent manner.
+
+So, what about HA in manufacturing assembly pipelines? The first approach in the 1940's was to reduce the service interruption in a production line just by converting internal machine components to external (we could compare this as a decoupling method for manufacturing, similar to decouple monoliths to distributed components), so to repair a machine in the *pipeline* would impact much less in the loss of service for the entire assembly line. ([Toyota started to apply this](http://artoflean.com/index.php/2010/02/15/set-up-reduction-in-toyota/) in the 1950's for machine setup, reducing times from hours to minutes, with a huge impact in the case or restarting any *pipeline* stage in a loss of service).
```

Comment:
"similar to *decoupling* monoliths to distributed components"
"so to repair a machine in the pipeline would result in much less loss of service"